### PR TITLE
API: break IncludeResources out of BaseController

### PR DESCRIPTION
**Pull request by AricStewart**

### CODE REVIEW

**Comment by rkaiser-dev** [comment]
I'm just curious, but what are we hoping to gain with this change specifically? I think in the end we will be removing this code anyway, in favor of using the service layer to manage this, right?

**Comment by AricStewart** [comment]
> I'm just curious, but what are we hoping to gain with this change specifically? I think in the end we will be removing this code anyway, in favor of using the service layer to manage this, right?

My goal is mostly soul soothing.  I am aghast as what I had let `BaseController` turn into.  So I have been trying to incrementally clean it up so that it is easier to parse what to salvage and what to discard.

I do not think we are going to discard all the include logic all at once.  So it seemed easy enough to pull it out so that it simplifies looking at `BaseController` My super secret goal was to trivialize down `__invoke` to the point where I could feel confident moving away from having the whole `buildResource` call and toward your goal of having each controller itself handle `__invoke`

**Comment by rkaiser-dev** [comment]
OK, I'll plan to take a more thorough look at this tomorrow. Thanks!

**Comment by rkaiser-dev** [comment]
I will take a look at this PR today. It may be best to hold off on merging this until after we've gotten the performance fix code out into production.

File: api/src/Controller/IncludeResource.php
Comment (rkaiser-dev) [inline api/src/Controller/IncludeResource.php:31]: I might be missing it, but where is this constructor being used?

**Review by rkaiser-dev — CHANGES_REQUESTED** [review]
There was a bit to wrap my head around here. I think overall this is trending in the right direction. 

It would be nice if we could break up `process` in `IncludeResource` so that it's using some smaller private functions. It's difficult to understand primarily because it is intended to be more generic, so I think smaller, descriptive functions would help.

File: api/src/Controller/BaseController.php
Comment (AricStewart) [reply api/src/Controller/BaseController.php:125]: This is because the new `IncludeResource` class need to access this int he base class and it is no longer in the same module so it no longer has access to protected members.  There is no friendship in PHP to allow unrelated classed to access protected members.

Comment (rkaiser-dev) [reply api/src/Controller/BaseController.php:125]: OK. I'll let this one slide since it's step 1 of moving `IncludeResource` out - typically stuff like this would be a code smell if trying to go the OOP route. It's effectively the same, but keeping this on `BaseController` and exposing a getter would allow access to this.

I think part of the danger with it being public vs using a getter is that if it's public, it could be changed by something else, where using a getter I think it would not be the case.

File: api/src/Controller/IncludeResource.php
Comment (AricStewart) [reply api/src/Controller/IncludeResource.php:31]: The constructor is used in each of the resources when it wants to signal something is included for example `Event/BaseEvent.php` line 115.

File: api/src/Controller/IncludeResource.php
Comment (AricStewart) [reply api/src/Controller/IncludeResource.php:42]: Yeah it is a pretty dense little recursive function.  When I get a some cycles I will try to help you unwind it.

Comment (AricStewart) [reply api/src/Controller/IncludeResource.php:42]:
```
    public function process(Request $request, Response $response, Container $container, array $params, array &$data, array $history): void
    {
        /* Verify this include field is in the resulting data, and make sure that the data is not null */
        if (in_array($this->field, array_keys($data), true) &&
            $data[$this->field] !== null) {
            /* Check if this resource class is in the history and if not add a blank history entry for this class */
            if (!array_key_exists($this->class, $history)) {
                $history[$this->class] = [];
            }
            /* check if in the class history this field has already been looked up. If so we do not continue */
            if (in_array($data[$this->field], $history[$this->class])) {
                return;
            }
            /* Build the parameters to call into the included resource */
            $newparams = $params;
            $newparams[$this->parameter] = $data[$this->field];
            $target = new $this->class($container);
            /* Call the buildResource for the included resource and if we crash just get out */
            try {
                $newdata = $target->buildResource($request, $response, $newparams)[1];
            } catch (Exception $e) {
                return;
            }
            /* Add this field to the history */
            $history[$this->class][] = $data[$this->field];
            /* Process all the includes for the newly included resource (Recursion) */
            IncludeResource::processIncludes($target->includes, $request, $response, $container, $params, $newdata, $history);
            /* Add all the response to the data for the resource */
            $data[$this->field] = $target->arrayResponse($request, $response, $newdata);
        }
```
Inlined comments because that was easiest for me to do.  I can include these comments if you feel like it makes it clearer.

Comment (rkaiser-dev) [reply api/src/Controller/IncludeResource.php:42]: The comments are helpful, and it confirms that I was on the right track. I think it would still be nice if we broke it into smaller functions though because it will give clarity through function names about what is being done and then we don't have to add comments to describe what is happening.

I'll let you decide which way to go with it. I'm including some example pseudocode for how it could be broken out into a smaller function just in case.

As a quick example, here's some pseudocode for what I mean:

```
public function process(Request $request, Response $response, Container $container, array $params, array &$data, array $history)
{
  if (hasIncludeData($data)) {
    initalizeHistory($history);
    if (hasExistingHistory($data, $history)) {
       return;
    }

    try {
       collectResourceData($request, $response, $container, $params, $data);
       $history[$this->class][] = $data[$this->field];
       processNextInclude($target->includes, $request, $response, $container, $params, $data, $newData, $history);
    } catch (Exception $e) {
       return;
    }
  }
}

private function hasIncludeData(array $data): boolean
{
  // Check if array has key and actually has data value
}

private function initalizeHistory(array $history): void
{
  // Check if the history exists and add it if not
}

private function hasExistingHistory(array $data, array $history): boolean
{
  // Check if the history already existed
}

private function collectResourceData(Request $request, Response $response, Container $container, array $params, array $data)
{
   // Include lines 50-54, do not include try/catch
}

private function processNextInclude ($includes, Request $request, Response $response, Container $container, array $params, array &$data, array $newdata, array $history): void
{
  IncludeResource::processIncludes($includes, $request, $response, $container, $params, $newdata, $history);
  $data[$this->field] = $target->arrayResponse($request, $response, $newdata);
}
```

**Review by rkaiser-dev — DISMISSED** [review]

**Comment by AricStewart** [comment]
Ok I met you half way. Keep some of the documentation and also broke a few function out.  We can revise more going forward if we want to.

**Pull request merged**
